Documentation/design-docs/host-startup-hook.md

```diff
@@ -0,0 +1,96 @@
+# Host startup hook
+
+For .NET Core 3+, we want to provide a low-level hook that allows
+injecting managed code to run before the main application's entry
+point. This hook will make it possible for the host to customize the
+behavior of managed applications during process launch, after they
+have been deployed.
+
+## Motivation
+
+This would allow hosting providers to define custom configuration and
+policy in managed code, including settings that potentially influence
+load behavior of the main entry point such as the
+`AssemblyLoadContext` behavior. The hook could be used to set up
+tracing or telemetry injection, to set up callbacks for handling
+Debug.Assert (if we make such an API available), or other
+environment-dependent behavior. The hook is separate from the entry
+point, so that user code doesn't need to be modified.
+
+## Proposed behavior
+
+The `DOTNET_STARTUP_HOOKS` environment variable can be used to specify
+a list of managed assemblies and type names that contain `public
+static void Initialize()` methods, which will be called in the order
+specified, before the `Main` entry point:
+
+```
+DOTNET_STARTUP_HOOKS=/path/to/StartupHook1.dll!StartupHookNamespace.StartupHookType1;/path/to/StartupHook2.dll!StartupHookNamespace.StartupHookType2
+```
+
+This variable is a list of absolute assembly paths, each with a type
+name following an exclamation mark. The list is delimited by the
+platform-specific path separator (`;` on Windows and `:` on Unix). It
+may not contain any empty entries or a trailing path separator.
```

Comment:
The code can presumably easily tolerate a trailing path separator

Reply:
Also, care is necessary if you use `;` separator. See https://github.com/dotnet/core-setup/issues/4141 for a break this already caused. `|` may be better and presumably works for Unix also. We would want to accept it for all the host configuration variables.

Reply:
> The code can presumably easily tolerate a trailing path separator

It could - see my comment above for the rationale behind the current choice. I'm fine with changing this if there's a good reason.

About the proper path separator - it's not clear that using `|` would better since it is allowed on unix. I think it's fine to specify that the path may not contain a `;` for this low-level hook, but am open to considering a more correct solution. It would probably take some more thought to figure out the right thing to do to handle weird but valid paths on different platforms.

Reply:
Paths with semicolons in seem to be very rare on Windows. In the linked issue it was there accidentally and the real problem was that it needed a debugger to figure that out. So if your error messages/tracing is good, I think semicolon is fine